scimma_admin/hopskotch_auth/views.py

```diff
@@ -74,5 +91,794 @@ def download(request):
+    return redirect(base_edit_url+"?topic_id="+str(topic.id))
+
+
+@login_required
+def edit_topic(request):
+    logger.info(f"User {request.user.username} ({request.user.email}) requested to edit topic ID " \
+                +request.GET.get("topic_id","<unset>")+" from "+request.META["REMOTE_ADDR"])
+    
+    if not "topic_id" in request.GET:
+        logger.info(f"Ignored request by user {request.user.username} ({request.user.email}) to " \
+                    +"edit a topic, as no topic ID was specified")
+        messages.error(request, "Missing topic ID.")
+        return redirect("index")
+    
+    try:
+        topic = KafkaTopic.objects.get(id=request.GET["topic_id"])
+    except ObjectDoesNotExist as dne:
+        logger.info(f"Ignored request by user {request.user.username} ({request.user.email}) to " \
+                    +"edit a topic, as no matching topic exists")
+        messages.error(request, "No such topic.")
+        return redirect("index")
+
+    owning_group = topic.owning_group
+    # only group owners and staff may use this page 
+    if not is_group_owner(request.user, owning_group.id) and not request.user.is_staff:
+        logger.info(f"Ignored request by user {request.user.username} ({request.user.email}) to " \
+                    +f"edit a topic owned by the group with ID {owning_group.id}, as that user is not a group owner or staff member")
+        messages.error(request, "Not authorized to manage that topic.")
+        return redirect("index")
+
+    permissions = []
+    for perm in topic.groupkafkapermission_set.all().select_related("principal"):
+        permissions.append({"group_name":perm.principal.name, "group_id":perm.principal.id, 
+                           "operation":perm.operation, "id":perm.id})
+
+    all_groups = Group.objects.all()
+    operations = KafkaOperation.__members__.keys()
+
+    return render(
+        request, "hopskotch_auth/edit_topic.html",
+        dict(topic=topic, owning_group=owning_group, permissions=permissions, all_groups=all_groups, operations=operations)
+        )
+
+
+@require_POST
+@login_required
+def set_topic_public_read_access(request):
+    logger.info(f"User {request.user.username} ({request.user.email}) requested to set topic ID " \
+                +request.POST.get("topic_id","<unset>")+" public read access to " \
+                +request.POST.get("public","<unset>")+" from "+request.META["REMOTE_ADDR"])
+
+    if not "topic_id" in request.POST:
+        logger.info(f"Ignored request by user {request.user.username} ({request.user.email}) to " \
+                    +"set public access to a topic, as no topic ID was specified")
+        messages.error(request, "Missing topic ID.")
+        return redirect("index")
+    if not "public" in request.POST:
+        logger.info(f"Ignored request by user {request.user.username} ({request.user.email}) to " \
+                    +"set public access to a topic, as no public status was specified")
```

Comment:
I'm working on making this change, although I'm a bit ambivilent about it: The reduction in length mostly comes from loss of detail in the error messages (e.g. removing the distinct messages for the end user and logged for staff).

Reply:
Right, I agree that it's not unambiguously good. Here's a possible improvement. It does away with the repetitive "operation" parameter, since we can just look up the name of the view from `request`, and it adds an `extra` dict of more data to include in the log.

```python
def fail_and_redirect(request, reason, redirect_to, extra=None):
    """
    request: The Django request being served.
    reason: A text string explaining the failure to the user.
    redirect_to: A URL to redirect to after the failure.
    extra (Optional): If set, additional information to log. Should be a dict of string:string pairs.
    """
    log_msg = (f"Ignored request by user {request.user.username} ({request.user.email}. " \
                + f"view={request.view}, reason={reason}")
    if extra is not None:
        log_msg += ", ".join([f"{k}={v}" for k, v in sorted(extra.items())])
    logger.info(log_msg)
    messages.error(request, reason)
    return redirect(redirect_to)
```

Then (including the switch to `exists()` over `try / except`):
```python
def set_topic_public_read_access(request):
    logger.info(f"User {request.user.username} ({request.user.email}) requested to set topic ID " \
                +request.POST.get("topic_id","<unset>")+" public read access to " \
                +request.POST.get("public","<unset>")+" from "+request.META["REMOTE_ADDR"])

    if not "topic_id" in request.POST:
        return fail_and_redirect(request, "Missing topic ID.", "index")
    if not "public" in request.POST:
        return fail_and_redirect(request, "Missing public status.", "index")

    if not KafkaTopic.objects.filter(id=request.POST["topic_id"]).exists():
        return fail_and_redirect(request, "No such topic.", "index")

    owning_group = topic.owning_group
    # only group owners and staff make a topic (not) publicly readable 
    if not is_group_owner(request.user, owning_group.id) and not request.user.is_staff:
        return fail_and_redirect(request, 
            "Not authorized to manage that topic.", 
            "index", 
            extra={"owning_group": owning_group.id, "topic": topic.id},
        )

    topic.publicly_readable = bool(request.POST["public"])
    topic.save()

    message="Made topic "+topic.name
    if topic.publicly_readable:
        message += " publicly readable"
    else:
        message += " not publicly readable"
    logger.info(message)
    messages.info(request, message)
    base_edit_url = reverse("edit_topic")
    return redirect(base_edit_url+"?topic_id="+str(topic.id))
```